app/src/main/java/org/xbmc/kore/ui/sections/remote/ShareHandlingFragment.java

```diff
@@ -0,0 +1,370 @@
+     * The Twitch app puts the url in the intent's EXTRA_TEXT along with
+     * some garbage.
+     */
+    static final ShareHandler TWITCH_APP = new ShareHandler() {
+        @Nullable
+        @Override
+        public String urlFrom(@Nullable String text) {
+            if (text == null) {
+                return null;
+            }
+            Matcher name = Pattern.compile(TWITCH_URL).matcher(text);
+            return name.find() ? TWITCH_PREFIX + name.group(1) + "/" : null;
+        }
+    };
+
+    private boolean handled;
+    private HostConnection connection;
+    private String playlistTag;
+
+    /**
+     * @param connection Share won't be handled when null.
+     */
+    public void connect(HostConnection connection) {
+        this.connection = connection;
+    }
+
+    /**
+     * The sibling PlaylistFragment should call this when visible so that
+     * it can be refreshed when the plugin url is added.
+     *
+     * @param tag The playlist fragment's tag. Use {@link Fragment#getTag()},
+     *            don't rely on FragmentPagerAdapter's implementation details.
+     */
+    public void setPlaylistTag(@Nullable String tag) {
+        playlistTag = tag;
+    }
+
+    @Override
+    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
+        super.onActivityCreated(savedInstanceState);
+        if (savedInstanceState != null) {
+            handled = savedInstanceState.getBoolean(KEY_HANDLED, false);
+        }
+    }
+
+    @Override
+    public void onStart() {
+        super.onStart();
+        if (handled || connection == null) {
+            return;
+        }
+        handled = true;
+
+        Intent intent = getActivity().getIntent();
+        String pluginUrl;
+        switch (intent.getAction()) {
+            case Intent.ACTION_SEND:
+                String extraText = intent.getStringExtra(Intent.EXTRA_TEXT);
+                pluginUrl = YOUTUBE_APP.urlFrom(extraText);
+                pluginUrl = pluginUrl != null ? pluginUrl : TWITCH_APP.urlFrom(extraText);
+                break;
+            case Intent.ACTION_VIEW:
+                pluginUrl = GENERIC.urlFrom(intent.getDataString());
+                break;
+            default:
+                return;
+        }
+
+        if (pluginUrl == null) {
+            say(R.string.error_share_video);
+            return;
+        }
+
+        final Handler h = new Handler();
+        final String file = pluginUrl;
+        isPlayingVideo(h).start(new OnFinish<Boolean>() {
+            @Override
+            public void got(Boolean isPlaying) {
+                if (isPlaying) {
+                    new Seq<>(enqueue(h, file))
+                            .then(hostNotify(h, getString(R.string.shared_video_added)))
+                            .start(refreshPlaylist());
+                } else {
+                    new Seq<>(clearPlaylist(h))
+                            .then(enqueue(h, file))
+                            .then(play(h))
```

Comment:
This is an incomplete form of the idiom from functional languages called the continuation monad. It's missing the type for `Task<U> from(T value)` and the corresponding handler for it in `Seq` because I have no use for it here. I think the usage is simple, just the implementation is tricky. If it has docs and tests it should be more usable.